src/traits-lowering-rules.md

```diff
@@ -198,38 +196,70 @@ in detail in the [section on associated types](./traits-associated-types.html),
 but reproduced here for reference:
 
 ```text
-  // Rule ProjectionEq-Normalize
-  //
-  // ProjectionEq can succeed by normalizing:
-  forall<Self, P1..Pn, Pn+1..Pm, U> {
-    ProjectionEq(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> = U) :-
-        Normalize(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> -> U)
-  }
+// Rule ProjectionEq-Normalize
+//
+// ProjectionEq can succeed by normalizing:
+forall<Self, P1..Pn, Pn+1..Pm, U> {
+  ProjectionEq(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> = U) :-
+      Normalize(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> -> U)
+}
+```
 
-  // Rule ProjectionEq-Skolemize
-  //
-  // ProjectionEq can succeed by skolemizing, see "associated type"
-  // chapter for more:
-  forall<Self, P1..Pn, Pn+1..Pm> {
-    ProjectionEq(
-      <Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> =
-        (Trait::AssocType)<Self, P1..Pn, Pn+1..Pm>
-    ) :-
-      // But only if the trait is implemented, and the conditions from
-      // the associated type are met as well:
-      Implemented(Self: Trait<P1..Pn>)
-      && WC1
-  }
+```text
+// Rule ProjectionEq-Skolemize
+//
+// ProjectionEq can succeed by skolemizing, see "associated type"
+// chapter for more:
+forall<Self, P1..Pn, Pn+1..Pm> {
+  ProjectionEq(
+    <Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> =
+    (Trait::AssocType)<Self, P1..Pn, Pn+1..Pm>
+  )
+}
 ```
 
 The next rule covers implied bounds for the projection. In particular,
-the `Bounds` declared on the associated type must be proven to hold to
-show that the impl is well-formed, and hence we can rely on them
+the `Bounds` declared on the associated type must have been proven to hold
+to show that the impl is well-formed, and hence we can rely on them
 elsewhere.
 
 ```text
-// XXX how exactly should we set this up? Have to be careful;
-// presumably this has to be a kind of `FromEnv` setup.
+// Rule Implied-Bound-From-AssocTy
+forall<Self, P1..Pn, Pn+1..Pm> {
```

Comment:
Could you add: « for each `Bound` in the set of `Bounds` » or something like this, to emphasize the fact that we output one `FromEnv` rule at a time